File: qt_ritual/test_assets/moqt/moqt_core/basic_class.cpp
#include "basic_class.h"

BasicClass::BasicClass(int x) : int_field(1), intReference_field(int_field) {
    m_foo = x;
    intPointerField = nullptr;
}

int BasicClass::foo() {
    return m_foo;
}

void BasicClass::setFoo(int foo) {
    m_foo = foo;
}


void BasicClass::updateFoo(UpdateTypes updateTypes) {
    if (updateTypes & Add2) {
        m_foo += 2;
    }
    if (updateTypes & Mul3) {
        m_foo *= 3;
    }
    if (updateTypes & Div5) {
        m_foo /= 5;
    }
}

QVector<int> BasicClass::get_vector_int() const {
    auto r = QVector<int>();
    r.append(1);
    r.append(3);
    r.append(5);
    return r;
}

QVector<BasicClassField> BasicClass::get_vector_class() const {
    auto r = QVector<BasicClassField>();
    BasicClassField value;
    value.set(2);
    r.append(value);
    value.set(4);
    r.append(value);
    value.set(6);
    r.append(value);
    return r;
}

BasicClass::operator int() {
    return 3;
}
BasicClass::operator QVector<int>() {
    auto r = QVector<int>();
    r.append(7);
    return r;
}

QFlags<BasicClass::UpdateType> operator|(BasicClass::UpdateType f1, BasicClass::UpdateType f2) {
    return QFlags<BasicClass::UpdateType>(f1 | f2);
}

void BasicClass::setRef(const int& value) {
}

File: qt_ritual/crate_templates/moqt_core/tests/basic_class.rs
use moqt_core::basic_class::UpdateType;
use moqt_core::{BasicClass, BasicClassField};
use std::ops::Deref;

#[test]
fn basic_class() {
    unsafe {
        let mut v = BasicClass::new(1);
        assert_eq!(v.foo(), 1);
        v.set_foo(5);
        assert_eq!(v.foo(), 5);

        assert_eq!(v.int_field(), 1);
        v.set_int_field(3);
        assert_eq!(v.int_field(), 3);

        assert!(v.int_pointer_field().is_null());
        let p = v.int_reference_field() as *mut i32;
        v.set_int_pointer_field(p);
        v.set_int_field(4);
        assert_eq!(v.int_pointer_field().deref(), &4);

        assert_eq!(v.int_reference_field(), &4);
        v.set_int_field(7);
        assert_eq!(v.int_reference_field(), &7);
        *v.int_reference_field() = 8;
        assert_eq!(v.int_field(), 8);

        // TODO: set_int_reference_field should have int arg
        v.set_int_reference_field(&mut 9);
        assert_eq!(v.int_field(), 9);

        assert_eq!(v.class_field().get(), 42);
        assert_eq!(v.class_field_mut().set(43), 42);
        assert_eq!(v.class_field().get(), 43);

        let c = BasicClassField::new();
        v.set_class_field(&c);
        assert_eq!(v.class_field().get(), 42);
        drop(c);
        assert_eq!(v.class_field().get(), 42);
    }
}

#[test]
fn nested_enum() {
    let x: UpdateType = UpdateType::Add2;
    assert_eq!(x.to_int(), 1);
    assert_eq!(UpdateType::Mul3.to_int(), 2);
    assert_eq!(UpdateType::Div5.to_int(), 4);

    unsafe {
        let mut v = BasicClass::new(1);
        v.set_foo(1);
        v.update_foo(UpdateType::Mul3.into());
        assert_eq!(v.foo(), 3);
        v.update_foo(UpdateType::Mul3 | UpdateType::Div5);
        assert_eq!(v.foo(), 1);
    }
}
